OutOfMemoryError in ExplodeJarTask

**Build scan link**
Private, can't share unfortunately.

**Plugin version**
`v1.13.1`, however no changes in `ExplodeJarTask` in ` main..v1.13.1`

**Gradle version**
`7.4.2`

**(Optional) Android Gradle Plugin (AGP) version**
Not applicable.

**`reason` output for bugs relating to incorrect advice**
Not applicable.

**Describe the bug**
When serializing the output of `ExplodeJarTask` to file, it sometimes throws an `OutOfMemoryError`:
```
Caused by: java.lang.OutOfMemoryError: Java heap space
at okio.Buffer.readString(Buffer.kt:313)
at okio.Buffer.readUtf8(Buffer.kt:297)
at com.squareup.moshi.JsonAdapter.toJson(JsonAdapter.java:94)
at com.autonomousapps.tasks.ExplodeJarTask$ExplodeJarWorkAction.execute(ExplodeJarTask.kt:126)
```

Specifically the call `explodedJars.toJson()` causes this. To give some context, `explodedJars.toJson().length = 5385952` in our case. The daemon is configured to run with 8Gb already.

**To Reproduce**
It's not happening deterministically unfortunately,  but having an insane number of dependencies sometimes triggers this.

**Expected behavior**
No OOM exception.

**Additional context**
From some initial investigation the entire `Set<ExplodedJar>` is first serialized to json and then written to file. Maybe it makes more sense to serialize the `ExplodedJar` one by one?

This is also similarly happening in `SynthesizeDependenciesTask` where the entire object is first converted to `String` and then written to the output file
```
Caused by: java.lang.OutOfMemoryError: Java heap space
at kotlin.io.TextStreamsKt.copyTo(ReadWrite.kt:125)
at kotlin.io.TextStreamsKt.copyTo$default(ReadWrite.kt:120)
at kotlin.io.TextStreamsKt.readText(ReadWrite.kt:107)
at kotlin.io.FilesKt__FileReadWriteKt.readText(FileReadWrite.kt:125)
at kotlin.io.FilesKt__FileReadWriteKt.readText$default(FileReadWrite.kt:125)
at com.autonomousapps.tasks.SynthesizeDependenciesTask$SynthesizeDependenciesWorkAction.execute(SynthesizeDependenciesTask.kt:234)
```

--- a/src/main/kotlin/com/autonomousapps/internal/utils/moshi.kt
+++ b/src/main/kotlin/com/autonomousapps/internal/utils/moshi.kt
@@ -10,6 +10,8 @@ import com.squareup.moshi.*
 import com.squareup.moshi.Types.newParameterizedType
 import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
 import dev.zacsweers.moshix.sealed.reflect.MetadataMoshiSealedJsonAdapterFactory
+import okio.buffer
+import okio.sink
 import java.io.File
 
 val MOSHI: Moshi by lazy {
@@ -104,6 +106,32 @@ inline fun <reified T> T.toPrettyString(withNulls: Boolean = false): String {
 
 inline fun <reified K, reified V> Map<K, V>.toPrettyString(withNulls: Boolean = false): String {
   return getJsonMapAdapter<K, V>(withNulls).indent("  ").toJson(this)
+}
+
+/**
+ * Buffers writes of the set to disk, using the indent to make the output human-readable.
+ * By default, the output is compacted.
+ *
+ * @param set The set to write to file
+ * @param indent The indent to control how the result is formatted
+ */
+inline fun <reified T> File.bufferWriteJsonSet(set: Set<T>, indent: String = "") {
+  JsonWriter.of(sink().buffer()).use { writer ->
+    getJsonSetAdapter<T>().indent(indent).toJson(writer, set)
+  }
+}
+
+/**
+ * Buffers writes of the set to disk, using the indent to make the output human-readable.
+ * By default, the output is compacted.
+ *
+ * @param set The set to write to file
+ * @param indent The indent to control how the result is formatted
+ */
+inline fun <reified T> File.bufferWriteJson(set: T, indent: String = "") {
+  JsonWriter.of(sink().buffer()).use { writer ->
+    getJsonAdapter<T>().indent(indent).toJson(writer, set)
+  }
 }
 
 @Suppress("unused")

--- a/src/main/kotlin/com/autonomousapps/tasks/ExplodeJarTask.kt
+++ b/src/main/kotlin/com/autonomousapps/tasks/ExplodeJarTask.kt
@@ -97,8 +97,8 @@ abstract class ExplodeJarTask @Inject constructor(
       ).explodedJars()
 
       // Write output to disk
-      outputFile.writeText(explodedJars.toJson())
-      outputPrettyFile.writeText(explodedJars.toPrettyString())
+      outputFile.bufferWriteJsonSet(explodedJars)
+      outputPrettyFile.bufferWriteJsonSet(explodedJars, "  ")
     }
   }
 }

--- a/src/main/kotlin/com/autonomousapps/tasks/SynthesizeDependenciesTask.kt
+++ b/src/main/kotlin/com/autonomousapps/tasks/SynthesizeDependenciesTask.kt
@@ -1,10 +1,7 @@
 package com.autonomousapps.tasks
 
 import com.autonomousapps.TASK_GROUP_DEP_INTERNAL
-import com.autonomousapps.internal.utils.fromJson
-import com.autonomousapps.internal.utils.fromJsonSet
-import com.autonomousapps.internal.utils.fromNullableJsonSet
-import com.autonomousapps.internal.utils.toJson
+import com.autonomousapps.internal.utils.*
 import com.autonomousapps.model.*
 import com.autonomousapps.model.intermediates.*
 import com.autonomousapps.services.InMemoryCache
@@ -182,10 +179,10 @@ abstract class SynthesizeDependenciesTask @Inject constructor(
         .map { it.build() }
         .forEach { dependency ->
           val coordinates = dependency.coordinates
-          outputDir.file(coordinates.toFileName()).get().asFile.writeText(dependency.toJson())
+          outputDir.file(coordinates.toFileName()).get().asFile.bufferWriteJson(dependency)
           if (coordinates is IncludedBuildCoordinates) {
             // Write the same information using the 'project' coordinates as later processing steps rely on it
-            outputDir.file(coordinates.resolvedProject.toFileName()).get().asFile.writeText(dependency.toJson())
+            outputDir.file(coordinates.resolvedProject.toFileName()).get().asFile.bufferWriteJson(dependency)
           }
         }
     }

